examples/utils/migration-tools/src/migrator/migrator.ts

```diff
@@ -7,395 +7,177 @@ import { TypedEventEmitter } from "@fluid-internal/client-utils";
-							exportedData,
-							migratedModel.version,
-						);
-					} catch {
-						// TODO: This implies that the contract is to throw if the data can't be transformed, which
-						// isn't great.  How should the dataTransformationCallback indicate failure?
-						this._events.emit("migrationNotSupported", acceptedMigration.newVersion);
-						this._migrationP = undefined;
-						return;
-					}
-				} else {
-					// We can't get the data into a format that we can import, give up.
-					this._events.emit("migrationNotSupported", acceptedMigration.newVersion);
-					this._migrationP = undefined;
-					return;
-				}
-				await migratedModel.importData(transformedData);
-
-				// Store the detached model for later use and retry scenarios
-				this._preparedDetachedModel = detachedContainer;
-			};
-
-			const completeTheMigration = async (): Promise<void> => {
-				assert(
-					this._preparedDetachedModel !== undefined,
-					"this._preparedDetachedModel should be defined",
-				);
-
-				// Volunteer to complete the migration.
-				let isAssigned: boolean;
-				try {
-					isAssigned = await this.currentMigrationTool.volunteerForMigration();
-				} catch {
-					// volunteerForMigration() will throw an error on disconnection. In this case, we should exit and
-					// re-enter the state machine which will wait until we reconnect.
-					// Note: while we wait to reconnect it is possible that another client will have already completed
-					// the migration.
-					assert(!this.connected, "We should be disconnected");
-					return;
-				}
-
-				if (this.currentMigrationTool.newContainerId !== undefined) {
-					// If newContainerId is already set, then another client already completed the migration.
-					return;
-				}
-
-				assert(isAssigned, "We should be assigned the migration task");
-
-				if (this._preparedModelId === undefined) {
-					this._preparedModelId = await this._preparedDetachedModel.attach();
-				}
-
-				// Check to make sure we still have the task assignment.
-				if (!this.currentMigrationTool.haveMigrationTask()) {
-					// Exit early if we lost the task assignment, we are most likely disconnected.
-					return;
-				}
-
-				await migrationTool.finalizeMigration(this._preparedModelId);
-
-				this.currentMigrationTool.completeMigrationTask();
-			};
-
-			// Prepare the detached model if we haven't already.
-			if (this._preparedDetachedModel === undefined) {
-				await prepareTheMigration();
+	private readonly performMigration = (): void => {
+		(async (): Promise<void> => {
+			const acceptedMigration = this.migrationTool.acceptedMigration;
+			assert(
+				acceptedMigration !== undefined,
+				"Expect an accepted migration before migration starts",
+			);
+			// Delay performing the migration until we are connected.  It's possible that we'll find the migration has already
+			// completed before we finish connecting, and in that case we want to avoid doing anything.
+			if (!this.connected) {
```

Comment:
What happens if we disconnect after this point? Do we need connection checks at more places in the flow than just here?

Reply:
For the most part it doesn't matter - both PactMap and CRC handle disconnect as we need them to, as long as we confirm the migration hasn't observably completed before attempting to write into them.

I have a TODO a little lower down about checking state more aggressively or using an abort signal, but this would mostly be (1) a performance optimization to realize we're done sooner if we lose the race and (2) avoid creating duplicate containers in more cases.  But no functional issue.  If we do that work, the abort signal might be driven partly by checking state upon reconnect.